src/librustc_ast_passes/ast_validation.rs

```diff
@@ -647,24 +647,61 @@ impl<'a> AstValidator<'a> {
             return;
         }
         // Find all generic argument coming after the first constraint...
-        let mut misplaced_args = Vec::new();
-        let mut first = None;
-        for arg in &data.args {
-            match (arg, first) {
-                (AngleBracketedArg::Arg(a), Some(_)) => misplaced_args.push(a.span()),
-                (AngleBracketedArg::Constraint(c), None) => first = Some(c.span),
-                (AngleBracketedArg::Arg(_), None) | (AngleBracketedArg::Constraint(_), Some(_)) => {
-                }
-            }
-        }
+        let constraint_spans = data
+            .args
+            .iter()
+            .filter_map(|arg| match arg {
+                AngleBracketedArg::Constraint(c) => Some(c.span),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+        let arg_spans = data
+            .args
+            .iter()
+            .filter_map(|arg| match arg {
+                AngleBracketedArg::Arg(a) => Some(a.span()),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+        let snippet_span = match &constraint_spans[..] {
+            [single] => *single,
+            [first, .., last] => first.to(*last),
+            [] => unreachable!(),
+        };
+        let removal_span = match &arg_spans[..] {
+            [first, ..] => snippet_span.until(*first),
+            [] => unreachable!(),
+        };
+        let sugg_span = match &arg_spans[..] {
+            [.., last] => last.shrink_to_hi(),
+            [] => unreachable!(),
+        };
+        let snippet = self.session.source_map().span_to_snippet(snippet_span).unwrap();
+        let constraint_len = constraint_spans.len();
         // ...and then error:
         self.err_handler()
             .struct_span_err(
-                misplaced_args.clone(),
+                arg_spans.clone(),
                 "generic arguments must come before the first constraint",
             )
-            .span_label(first.unwrap(), "the first constraint is provided here")
-            .span_labels(misplaced_args, "generic argument")
+            .span_labels(
+                constraint_spans,
+                &format!(
+                    "the constraint{} {} provided here",
+                    pluralize!(constraint_len),
+                    if constraint_len == 1 { "is" } else { "are" }
+                ),
```

Comment:
```suggestion
                "constraint"
```

Feels nicer (and simpler in terms of code) to just label what the syntactic categories here are and then the suggestion below is responsible for telling the user what to do about it.